Modgobi 25 b

mtraneis opened the pull request:
Created the config data json schema.  Schema covers the cases of recursive mapping and referencing another map (that has been evaluated).

id-jenkins commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -23,6 +23,8 @@
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 
+import javax.xml.crypto.Data;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused import 'javax.xml.crypto.Data'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

id-jenkins commented:
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

id-jenkins commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
+
+    String from = dataSource.getFrom();
+
+    DataSource dataSrc = DataSource.builder()
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "dataSrc". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)

craigmcnally commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -140,13 +140,7 @@ public PostGobiOrdersHelper(HttpClientInterface httpClient,
-            return CompletableFuture.completedFuture(null);
-          }
-        })
+        .withTranslation(this::getPurchaseOptionCode)
```
This should probably be pulled out of here... I never intended for it to get checked-in to begin with.  I special-cased it just to put "something" in this field.  [MODGOBI-11](https://issues.folio.org/browse/MODGOBI-11) will need to implement a Mapper function that parses the PurchaseOption Code and return the numeric user limit.  This translation method should be specified just like other translators.

mtraneis replied:
I have something to that effect, though not exactly right yet.  I extracted that logic into the method "getPurchaseOptionCode" and listed as an option for translation in the json schema.  So, is this work part of this user story? I only touched it because it had to do with the schema.

craigmcnally replied:
You can just leave it as-is for now and well tackle it in MODGOBI-11

craigmcnally commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -159,6 +153,14 @@ public PostGobiOrdersHelper(HttpClientInterface httpClient,
     return future;
   }
 
+  public CompletableFuture<Integer> getPurchaseOptionCode(Object s){
```
See above.  The implementation of this will eventually be replaced with something that parses the purchase option code and returns the numeric user limit

craigmcnally commented on src/test/java/org/folio/rest/impl/PostGobiOrdersHelperTest.java:
```diff
@@ -355,9 +361,54 @@ public final void testLookupOrderMappings(TestContext context) throws Exception
   }
+
+  @Test
+  public final void testextractOrderMappingWithOtherField(TestContext context) throws Exception {
```
Fix method name case... testExtract

craigmcnally replied:
Also, this test case probably doesn't belong in this test file since it doesn't pertain to PostGobiOrdersHelper.java.

craigmcnally commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
-            logger.error("Combinator method not found: " + combinator, e);
+            return (String) combinatorMethod.invoke(null, data);
+          } catch (Exception e) {
+            logger.error("Unable to invoke combinator method: " + combinator, e);
```
it might be helpful to log additional context... e.g. the full mapping or at least the field

craigmcnally commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
-            logger.error("Translation method not found: " + translation, e);
+            return (CompletableFuture<Object>) translationMethod.invoke(null, data);
+          } catch (Exception e) {
+            logger.error("Unable to invoke translation method: " + translation, e);
```
it might be helpful to log additional context... e.g. the full mapping or at least the field

craigmcnally commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
+
+    if (dataSource.getDefault() != null) {
+      defaultValue = dataSource.getDefault();
+    }
```
inconsistent formatting... else should be on same line as curly brace

craigmcnally commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
+      String otherField = dataSource.getFromOtherField().value();
+      defaultValue = map.get(Field.valueOf(otherField));
+    }
+    else if(dataSource.getDefaultMapping() != null) {
```
see formatting comment above

craigmcnally commented on src/main/resources/mapping.json:
```diff
@@ -36,15 +36,24 @@
+          "$ref" : "#",
+          "description" : "default value found from the mapping of another node"
+        },
+        "fromOtherField" : {  //in the config json, the order of dataSource is important as this field refers to previous dataSource objects.
```
I think the comment makes this invalid json.  Maybe move this to the description?

craigmcnally commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +91,74 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
+    NodeCombinator nc = null;
+    if (combinator != null) {
+      try {
+        Method combinatorMethod = Mapper.class.getMethod(combinator.toString(), NodeList.class);
```
I thought we discussed not using reflection... not a huge deal though

id-jenkins commented on src/main/java/org/folio/gobi/HelperUtils.java:
```diff
@@ -90,54 +89,69 @@ public static String encodeValue(String value) throws UnsupportedEncodingExcepti
+      defaultValue = extractOrderMapping(dataSource.getDefaultMapping().getDataSource(), map);
+    }
+
+    DataSource dataSrc = DataSource.builder()
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "dataSrc". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)

craigmcnally replied:
Can you resolve this please?

id-jenkins commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

mtraneis commented:
Addressed all issues raised in the last commit except for the getCustomOrderCode lookup and calling a translation method using reflection.  The former will go into MODGOBI-11 and the latter requires refactoring, which deserves its own user story.
